Feature/20 Registration Service initial ADR

Izzzu opened the pull request:
Registration Service ADR specifying the architecture for the MVD. 
This document is derived from the [Registration Service](https://github.com/Metaform/mvd/blob/main/registration-service/registration-service-tech-spec.md) Specification and focuses on what will be implemented for the MVD.

closes #20

cpeeyush commented on docker-compose.yml:
```diff
@@ -8,8 +8,6 @@ services:
-      NODES_JSON_DIR: /resources/registry
     ports:
       - "8181:8181"
       - "5005:5005"
```
volumes can be removed as well.

Adding comment here it is not reflected in diff.

cpeeyush commented on extensions/participant-store-memory/build.gradle.kts:
```diff
@@ -0,0 +1,20 @@
+    `java-library`
+}
+
+val rsApi: String by project
```
unused.
```suggestion
```

cpeeyush reviewed (approved):
Approved with small suggestions.

algattik commented:
Moved to new repo

Izzzu committed:
ADR

Izzzu committed:
ADR

Izzzu committed:
ADR

Izzzu committed:
ADR

Izzzu committed:
ADR

Izzzu committed:
ADR

Izzzu committed:
ADR

algattik commented on docs/developer/decision-records/2022-06-13-registration-service/README.md:
```diff
@@ -0,0 +1,137 @@
+# Registration Service
```
could we make the folder name more specific?

algattik commented on docs/developer/decision-records/2022-06-13-registration-service/README.md:
```diff
@@ -0,0 +1,137 @@
+
+In simple scenarios, enrollment could be fast and fully automated. However, in advanced scenarios, enrollment policies could require interactions with external systems, and even manual processes. Therefore, it is implemented asynchronously.
+
+There could be different "types" of onboarding, e.g. onboarding a participant or a credential issuer, so the architecture has to support that.
```
```suggestion
```

Delete this until we can make the point more specific

algattik commented on docs/developer/decision-records/2022-06-13-registration-service/dataspace-enrollment.puml:
```diff
@@ -0,0 +1,27 @@
+box "Dataspace Authority" #LightYellow
+Participant "Dataspace\nDID Document" as DID0
+Participant "DA\nIdentity Hub" as IDHUB0
+Participant "DA API" as DA
```
```suggestion
Participant "Registration Service" as DA
```

algattik commented on docs/developer/decision-records/2022-06-13-registration-service/dataspace-enrollment.puml:
```diff
@@ -0,0 +1,27 @@
+DA -> Registry: Store membership
+DA -> DA: Sign membership certificate
+DA -> IDHUB1 ++: signed membership credential:\n - Hub Token
+IDHUB1 -> IDHUB1: Store membership credential
```
2 vertical activation boxes are not closed in the bottom

algattik commented on docs/developer/decision-records/2022-06-13-registration-service/README.md:
```diff
@@ -0,0 +1,137 @@
+Responsibilities of the Dataspace Authority in MVD version:
+- Establishes a DID defining the Dataspace.
+- Manages policies defining who is allowed to participate in the dataspace.
+- Manages enrollment process of the Dataspace participants.
```
enrollment/onboarding seem to be used as synonyms. Make uniform

Izzzu replied:
Changed to enrollment

algattik commented on docs/developer/decision-records/2022-06-13-registration-service/README.md:
```diff
@@ -0,0 +1,137 @@
+
+![dataspace-enrollment](dataspace-enrollment.png)
+
+As a precondition to this operation Company1 knows the DID URL of the Dataspace it intends to join.
```
Move to a "Preconditions" section and add other necessary preconditions (e.g. identity hub for company1 deployed). See https://github.com/agera-edc/MinimumViableDataspaceOld/blob/ad8ff7b97905f784a4d9107d129cc3f1cd3102c2/docs/developer/decision-records/2022-05-10-dataspace-communication/README.md?plain=1#L61

algattik commented on docs/developer/decision-records/2022-06-13-registration-service/README.md:
```diff
@@ -0,0 +1,137 @@
+
+A Client for Company1 initiates the enrollment process by resolving and contacting the enrollment API endpoint for the Dataspace Authority. The client could be e.g. a CLI utility.
+
+The Dataspace Authority enrollment service obtains Verifiable Credentials from Company1 to determine whether it meets enrollment policies. The enrollment service then issues a Verifiable Credential that establishes membership and pushes it to Company 1's Identity Hub, and stores membership and certificate information.
```
"DA", "Dataspace Authority enrollment service", "Registration Service", "Participant registry", "Dataspace registry" seem to be used as synonyms. Make uniform

Izzzu replied:
As I understood the Registration Service in MVD version is a Dataspace Authority. That's why sometimes in the doc it's mentioned as Reg Service and sometimes as Dataspace Authority for more generic meaning. 
I will try to make it more understandable

algattik commented on docs/developer/decision-records/2022-06-13-registration-service/README.md:
```diff
@@ -0,0 +1,137 @@
+3. The DA authenticates the request...
+4. ... and retrieves credentials from Company1's Identity Hub.
+5. The DA authorizes the request by applying the Dataspace enrollment policy on the obtained Verifiable Credentials.
+6. The DA stores membership information in its registry. At the very least, this includes Company 1's DID URL.
```
```suggestion
6. The DA stores membership information in its registry. This includes Company 1's DID URL.
```

algattik commented on docs/developer/decision-records/2022-06-13-registration-service/README.md:
```diff
@@ -0,0 +1,137 @@
+     * For the MVD this is a JWT that contains the candidate's DID URI, signed with its private key.
+     * In future versions this could also contain other tokens.
+     */
+    private String identityToken;
```
this is in the JWS Authentication header, therefore the request can be empty.

algattik commented on docs/developer/decision-records/2022-06-13-registration-service/README.md:
````diff
@@ -0,0 +1,137 @@
+}
+```
+
+Upon receiving an OnboardingRequest, the registration service creates a participant record, that contains all information about a participant.
````
I think we should avoid mixing architecture and code design in one ADR. It's fine to have 'conceptual level' domain design here, but we shouldn't document Java types especially if they are the internals (as opposed to the API interface) of the service.

At this stage I believe we only need DID URI (which is also the unique identifier) and ParticipantStatus

algattik commented on docs/developer/decision-records/2022-06-13-registration-service/README.md:
```diff
@@ -0,0 +1,137 @@
+
+#### Participants
+
+1. Company1, a Dataspace Participant with an EDC application that wants to discover IDS endpoints (for instance, in order to list contract offers)
```
```suggestion
1. Company1, a Dataspace Participant with a Dataspace Connector (e.g. EDC application) that wants to discover IDS endpoints (e.g. in order to list contract offers)
```

algattik commented on docs/developer/decision-records/2022-06-13-registration-service/README.md:
````diff
@@ -0,0 +1,137 @@
+}
+```
+
+### List participants flow
````
Make format consistent with previous heading at same level

algattik commented on docs/developer/decision-records/2022-06-13-registration-service/README.md:
```diff
@@ -0,0 +1,137 @@
+2. The client needs access to the Company1 Private Key to sign a JWS. The client sends an HTTP request to the Dataspace Authority enrollment endpoint. The 
+   request is accepted for asynchronous processing.
+3. The DA authenticates the request...
+4. ... and retrieves credentials from Company1's Identity Hub.
```
```suggestion
4. ... and retrieves Verifiable Presentations from Company1's Identity Hub.
```

algattik commented on docs/developer/decision-records/2022-06-13-registration-service/README.md:
```diff
@@ -0,0 +1,137 @@
+   request is accepted for asynchronous processing.
+3. The DA authenticates the request...
+4. ... and retrieves credentials from Company1's Identity Hub.
+5. The DA authorizes the request by applying the Dataspace enrollment policy on the obtained Verifiable Credentials.
```
```suggestion
5. The DA authorizes the request by applying the Dataspace enrollment policy on the obtained Verifiable Presentations.
```

algattik commented on docs/developer/decision-records/2022-06-13-registration-service/README.md:
```diff
@@ -0,0 +1,137 @@
+
+#### Flow sequence
+
+![dataspace-enrollment](dataspace-enrollment.png)
```
the diagram shows a sync version of the flow. Update to account for async behavior? E.g. participant is stored in registry immediately, and 6 is actually a state update

"Participant registry" should probably be renamed to "Participant store" or similar

algattik commented on docs/developer/decision-records/2022-06-13-registration-service/README.md:
```diff
@@ -0,0 +1,137 @@
+## References
+
+- [Identity Hub in MVD](https://github.com/agera-edc/IdentityHubFork/blob/main/docs/developer/decision-records/2022-06-08-identity-hub/README.md)
+- [Technical specification of Registration Service](https://github.com/Metaform/mvd/blob/main/registration-service/registration-service-tech-spec.md)
```
```suggestion
```
Remove links to personal repos; this doc should be self-sufficient

algattik commented on docs/developer/decision-records/2022-06-13-registration-service/README.md:
```diff
@@ -0,0 +1,137 @@
+
+## References
+
+- [Identity Hub in MVD](https://github.com/agera-edc/IdentityHubFork/blob/main/docs/developer/decision-records/2022-06-08-identity-hub/README.md)
```
```suggestion
- [Identity Hub in MVD](https://github.com/agera-edc/IdentityHub/blob/main/docs/developer/decision-records/2022-06-08-identity-hub/README.md)
```

algattik committed:
Updates to introduction section

algattik commented:
@Izzzu I've pushed some suggested text improvements

Izzzu commented:
@algattik @cpeeyush I'm closing this PR as it's moved to MVD. I addressed comments mentioned here.
PR in MVD: https://github.com/agera-edc/MinimumViableDataspace/pull/143

I will open a new PR in Registration Service containing ADR with information related only to RS that are already included in the repository (API, in-memory storage)
